signxml/__init__.py

```diff
@@ -389,9 +395,10 @@ def sign(self, method=methods.enveloped, algorithm="rsa-sha256", key=None, passp
                 signature = long_to_bytes(r).rjust(32, b"\0") + long_to_bytes(s).rjust(32, b"\0")
 
             signature_value.text = ensure_str(b64encode(signature))
-
             key_info = SubElement(self.sig_root, ds_tag("KeyInfo"))
-            if cert_chain is None:
+            if key_name is not None:
+                SubElement(key_info, ds_tag("KeyName")).text = key_name
+            elif cert_chain is None:
```

Comment:
Hi Andrey, you are right, they are not mutually exclusive, according to
scheme this is definition:

 Schema Definition:

   <element name="KeyInfo" type="ds:KeyInfoType"/>
   <complexType name="KeyInfoType" mixed="true">
     <choice maxOccurs="unbounded">
       <element ref="ds:KeyName"/>
       <element ref="ds:KeyValue"/>
       <element ref="ds:RetrievalMethod"/>
       <element ref="ds:X509Data"/>
       <element ref="ds:PGPData"/>
       <element ref="ds:SPKIData"/>
       <element ref="ds:MgmtData"/>
       <any processContents="lax" namespace="##other"/>
       <!-- (1,1) elements from (0,unbounded) namespaces -->
     </choice>
     <attribute name="Id" type="ID" use="optional"/>
   </complexType>


But now the use-case is important: I do not want always to include
KeyInfo/KeyValue/X509Data
How to express for that ?

Would it be better to pass list of KeyInfo options to sign method ?

for example syntax:

...sign(... KeyInfo = { 'KeyName':<string>, 'KeyValue': <boolean>,
'X509Data': <boolean>} )

Sample call:
...sign(... KeyInfo = { 'KeyName':'MyName', 'KeyValue': False, 'X509Data':
False} )

What would be your preference ?

rgds Harry

2015-09-11 14:36 GMT+02:00 Andrey Kislyuk <notifications@github.com>:

> In signxml/__init__.py
> <https://github.com/kislyuk/signxml/pull/24#discussion_r39266448>:
>
> >              key_info = SubElement(self.sig_root, ds_tag("KeyInfo"))
> > -            if cert_chain is None:
> > +            if key_name is not None:
> > +                SubElement(key_info, ds_tag("KeyName")).text = key_name
> > +            elif cert_chain is None:
>
> These conditions are not mutually exclusive. This has to be an if, not an
> elif, if I understand correctly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kislyuk/signxml/pull/24/files#r39266448>.
>